source/common/http/http1/parser.h

```diff
@@ -0,0 +1,183 @@
+   * trailers are signaled via onMessageCompleteBase().
+   * @return An error status or a ParserStatus.
+   */
+  virtual Envoy::StatusOr<ParserStatus> onHeadersComplete() PURE;
+
+  /**
+   * Called when body data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length
+   */
+  virtual void bufferBody(const char* data, size_t length) PURE;
+
+  /**
+   * Called when the HTTP message has completed parsing.
+   * @return An error status or a ParserStatus.
+   */
+  virtual StatusOr<ParserStatus> onMessageComplete() PURE;
+
+  /**
+   * Called when accepting a chunk header.
+   */
+  virtual void onChunkHeader(bool) PURE;
+
+  virtual int setAndCheckCallbackStatus(Status&& status) PURE;
+  virtual int setAndCheckCallbackStatusOr(Envoy::StatusOr<ParserStatus>&& statusor) PURE;
+};
+
+class Parser {
+public:
+  // Struct containing the return value from parser execution.
+  struct RcVal {
+    // Number of parsed bytes.
+    size_t nread;
+    // Integer error from parser indicating return code.
+    int rc;
+  };
+  virtual ~Parser() = default;
+
+  // Executes the parser.
+  // @return an RcVal containing the number of parsed bytes and return code.
+  virtual RcVal execute(const char* slice, int len) PURE;
+
+  // Unpauses the parser.
+  virtual void resume() PURE;
+
+  // Pauses the parser and returns a status indicating pause.
+  virtual ParserStatus pause() PURE;
+
+  // Returns a parser status representing the errno value from the parser.
+  virtual ParserStatus getStatus() PURE;
+
+  // Returns an integer representing the status code stored in the parser structure. For responses
+  // only.
+  virtual uint16_t statusCode() const PURE;
+
+  // Returns an integer representing the HTTP major version.
+  virtual int httpMajor() const PURE;
+
+  // Returns an integer representing the HTTP minor version.
+  virtual int httpMinor() const PURE;
+
+  // Returns the number of bytes in the body. -1 if no Content-Length header
+  virtual uint64_t contentLength() const PURE;
+
+  // Returns parser flags (e.g. chunked).
+  virtual int flags() const PURE;
+
+  // Returns an integer representing the method. For requests only.
+  virtual uint16_t method() const PURE;
+
+  // Returns a textual representation of the method. For requests only.
+  virtual const char* methodName() const PURE;
+
+  // Returns a textual representation of the latest return error.
+  virtual const char* errnoName() PURE;
+
+  // Returns a textual representation of the given return code.
+  virtual const char* errnoName(int rc) const PURE;
+
+  // Returns whether the Transfer-Encoding header is present.
+  virtual int usesTransferEncoding() const PURE;
+
+  // Returns whether the Content-Length header is present.
+  virtual bool seenContentLength() const PURE;
+
+  // Tells the parser that the Content-Length header is present.
+  virtual void setSeenContentLength(bool val) PURE;
```

Comment:
is this needed?

Reply:
This is because the new parser doesn't have an "unset" content-length like http_parser. But I will remove this for now, and add that in with the new impl.